### PR TITLE
added information when base container is missing

### PR DESCRIPTION
**Pull request by svenwillger**
Please review changes

### CODE REVIEW

File: workflow/rules/functions.smk [hunk 1]
```diff
@@ -17,7 +17,13 @@ def _get_container_metadata(refcon_name=None):
     else:
         refcon_md = [(key.split('_')[-1], metadata) for key, metadata in config.items() if key.startswith('metadata_')]
     if not refcon_md:
-        raise ValueError('No "metadata_" keys found in config. Did you forget to load a reference container config?')
+        base_image_name = [(key.split('_')[-1], metadata) for key, metadata in config.items() if key.startswith('metadata_')]
+        base_image_name = [(base_image, metadata) for base_image, metadata in base_image_name if base_image in select_names]
```
Comment (ptrebert) [inline workflow/rules/functions.smk]: `select_names` may be undefined in that scope

File: workflow/rules/functions.smk [hunk 2]
```diff
@@ -17,7 +17,13 @@ def _get_container_metadata(refcon_name=None):
     else:
         refcon_md = [(key.split('_')[-1], metadata) for key, metadata in config.items() if key.startswith('metadata_')]
     if not refcon_md:
-        raise ValueError('No "metadata_" keys found in config. Did you forget to load a reference container config?')
+        base_image_name = [(key.split('_')[-1], metadata) for key, metadata in config.items() if key.startswith('metadata_')]
```
Comment (ptrebert) [inline workflow/rules/functions.smk]: if not `metadata_` keys are in the config (no config loaded), that list should be empty, and, consequently, the following lines won't reach the desired state

File: workflow/rules/functions.smk [hunk 3]
```diff
@@ -17,7 +17,13 @@ def _get_container_metadata(refcon_name=None):
...
+        base_image_name = [(base_image, metadata) for base_image, metadata in base_image_name if base_image in select_names]
+        base_container_path = pl.Path(f'container/{base_image_name}.sif')
```
Comment (ptrebert) [inline workflow/rules/functions.smk]: `base_image_name` is a list here; that f-string formatting won't work as far as I can tell

File: workflow/rules/functions.smk [hunk 4]
```diff
@@ -65,7 +71,7 @@ def create_refcon_base_map():
 
 def _find_base_def_file(base_name):
 
-    repo_path = pl.Path(workflow.basedir).absolute()
+    repo_path = pl.Path(workflow.basedir).parent.resolve(strict=True)
```
Comment (ptrebert) [inline workflow/rules/functions.smk:97]: please add a `TODO` for the future here: this line needs to be changed when we adapt the refcon repo to the Snakemake template, and use a constant from the template

File: workflow/rules/functions.smk [hunk 5]
```diff
@@ -380,6 +386,10 @@ def build_filename_constraints(which):
 def select_container_readme(wildcards):
 
     _, refcon_md = _get_container_metadata(wildcards.rc_name_version)[0]
+    if 'base' in {wildcards.rc_name_version}:
+        base_container_path = pl.Path(f'container/{wildcards.rc_name_version}.sif')
+        if not base_container_path.is_file():
+            raise ValueError('No base container found. Please follow instructions above to create base container')
```
Comment (ptrebert) [inline workflow/rules/functions.smk]: please make this even more explicit:
"...instructions above (see: ACTION REQUIRED) to create..."

**Review by ptrebert — CHANGES_REQUESTED** [review]
as just discussed via WebEx, please also make sure to test the changes under different conditions (no base image, no container config, etc.)

**Comment by svenwillger** [comment]
All the changes I made before to the `functions.smk`module and we discussed didn't solve the problem, that's why I'm not replying to the raised points in the change request. 
I went thoroughly through the workflow and the reason for the misleading error message ('No "metadata_" keys found in config. Did you forget to load a reference container config?') was that the `functions.smk` loops several times through the function "_get_container_metadata" and creates a "new" 'refcon_md' every time another function refers to "_get_container_metadata" and at some point it arrives with the name of the desired container (e.g. broad-hg38...) in "refcon_name", which triggers the first if condition "if refcon_name is not None"  but during that loop in line 12 'select names' only expects "base" and "base_v0" resulting in an error that leaves 'refcon_md' empty and therefore triggers the error message and then stops the script.

I was able to find an easy solution to the whole issue!

The function "collect_base_images" starting in line 78 already deals with the case what to do when there is no file 'container/base_v0.sif' and displays the instructions what to do, BUT it was missing a stop sign in the function if that problem occurs.

I just added the last line to the section with the instructions:
```
info_msg = f'\nACTION REQUIRED --- missing base container detected: {base_image}.sif\n'
info_msg += f'Please run the following build command in the "container" folder:\n'
info_msg += f'$ cd {workflow_dir}/container\n'
info_msg += f'$ sudo singularity build {base_image}.sif {base_image}.def\n\n'
info_msg += '(Restart the pipeline afterwards to continue building the reference container)\n\n'
sys.stderr.write(info_msg)
raise ValueError('Container is missing. Please follow the instructions above!!')
```
This stops the script right there and you can see the instructions directly above the error message and includes now the line that the pipeline needs to be started again to build the actual reference container.

The same is true for the actual reference container image. Here the problem is in  `packaging.smk`in the rule "notify user".
The error that was originally thrown was a missing input and after 5 sec waiting it stops. I now also added the ValueError so that the script stops immediately and displays that the instructions what to do are just above the error message.

I tested the new workflow several times, with the broad and ncbi configs and with several possible cases (all correct input provided, no sif file(s), no "metadata_" in the yaml, no yaml) and the workflow behaves in all cases as expected.

I pushed the new commit to all and rebased feature onto dev.  

The latest commit (hash ID `20751`) was submitted with the message "fix when base container is missing" and should fix the issue. Minor changes we discussed, like adding the `TODO`regarding the parent folder and changing `singularity:` to `container:` will be submitted separately.

**Comment by ptrebert** [comment]
I have changed the funtion `collect_base_images` again to restore the desired intent:
1) note the comment: iterate through configs and collect *all* base images; throwing an error immediately could lead to the annoying behavior of reporting errors one by one after each restart
2) however, starting the build process for all available base images, and failing for the missing ones may also not be perfect, as base images should be quick to build and add. Consequently, the behavior of the pipeline is now changed to collect all errors, and stops if there is at least one missing base image.

File: workflow/rules/functions.smk (hunk 4 above)
Comment (ptrebert) [reply workflow/rules/functions.smk:97]: TODO added

**Review by ptrebert — APPROVED** [review]

**Comment by ptrebert** [comment]
Rebasing onto dev and merging done

**Pull request merged**